libsolidity/interface/OptimiserSettings.h

```diff
@@ -155,6 +155,8 @@ struct OptimiserSettings
 	/// This specifies an estimate on how often each opcode in this assembly will be executed,
 	/// i.e. use a small value to optimise for size and a large value to optimise for runtime gas usage.
 	size_t expectedExecutionsPerDeployment = 200;
+	/// Flag reflecting whether optimizer is enabled.
+	bool enabled = false;
```

Comment:
I see there has been some discussion about this here: https://github.com/ethereum/solidity/pull/12834#discussion_r869172773 - adding this flag still strikes me as odd here.
What happens if you *always* run the optimizer (if only with minimal optimiser settings)?
I haven't *really* looked at the import/export test script, but does it actually test that bytecode matches between direct compilation and compilation after export+import (since it should)?

So
```
solc --asm-json test.sol > test.asm
solc --bin --import-asm-json test.asm
```
should produce the same bytecode as
```
solc --bin test.sol
```

And also ideally all of these:
```
solc --optimize --asm-json test.sol > test.asm
solc --bin --import-asm-json test.asm
```
and
```
solc --optimize --asm-json test.sol > test.asm
solc --optimize --bin --import-asm-json test.asm
```
and
```
solc --asm-json test.sol > test.asm
solc --optimize --bin --import-asm-json test.asm
```
should produce the same output as

```
solc --optimize --bin test.sol
```

Maybe we don't need all of that, but can you tell which of them work and how this relates to the optimizer setting here?

Reply:
Sorry for the late answer!

> What happens if you always run the optimizer (if only with minimal optimiser settings)?

If we would run always the optimiser, we would never be able to have matching bytecode. Because we have nothing we could compare against.
We will not be able to get the same byte code for the normal solidity based optimisation e.g. `solc —optimize —bin` compared with an optimisation run by using the asm-import e.g. `solc --optimize --bin --import-asm-json test.asm`.  Both use totally different optimiser steps, see https://github.com/ethereum/solidity/pull/12704#issuecomment-1067363681.

I can basically only see, that the resulting byte-code is becoming shorter in most cases, if optimizer was turned on for the asm-import. But I cannot compare it with anything against, because `CompilerStack` is using `AssemblyStack::optimize()`, where the import uses `Assembly::optimise(..)`, `Assembly::optimiseInternal(..)`.  The final result of the different optimisation steps are different. Thats why the import/export tests are not explicitly testing the optimisation behaviour.

Somehow I would tend now to remove the optimisation completely from the EVM Assembly import. Maybe it is better just to generate the code exactly how it was defined by the EVM Assembly - without any alteration - without any optimisation.

Reply:
Also another note to `solc --asm-json test.sol > test.asm` - sadly - at least for now - is not returning well-formed json like. `solc --asm-json` does not return well formed json, only well formed json fragments. E.g.:
```
$ solc/solc --optimize --asm-json test.sol
======= /Users/alex/git/solidity/test/libsolidity/semanticTests/constructor_inheritance_init_order.sol:A =======
EVM assembly:
{".code":[{"begin":0,"end":135,"name":"PUSH","source":0,"value":"80"},{"begin":0,"end":135,"name":"PUSH","source":0,"value":"40"},{"begin":0,"end":135,"name":
...
":0},{"begin":71,"end":133,"name":"SWAP1","source":0},{"begin":71,"end":133,"name":"RETURN","source":0}]}},"sourceList":["/Users/alex/git/solidity/test/libsolidity/semanticTests/constructor_inheritance_init_order.sol","#utility.yul"]}

======= /Users/alex/git/solidity/test/libsolidity/semanticTests/constructor_inheritance_init_order.sol:B =======
EVM assembly:
{".code":[{"begin":136,"end":180,"name":"PUSH","source":0,"value":"80"},{"begin":136,"end":180,"name":"PUSH","source":0,"value":"40"},{"begin":136,"end":180,"
...
{"begin":158,"end":177,"name":"JUMP","source":0}]}},"sourceList":["/Users/alex/git/solidity/test/libsolidity/semanticTests/constructor_inheritance_init_order.sol","#utility.yul"]}
```
JSON data is returned between potentially multiple `======= file:contractName =======` delimiters.
I know this would be a breaking change, but I guess we should change that to return well-formed json - even if run on the console - right?

Reply:
Regarding the ```====...====`` delimiters and piping: yeah, that's inconvenient, but we can still ignore that for now.

However, I don't think https://github.com/ethereum/solidity/pull/12704#issuecomment-1067363681 is correct, i.e. I don't think that's what's happening: the ``AssemblyStack`` (now ``YulStack``) optimizer is the Yul optimizer that's completely orthogonal to assembly import/export, since it happens *before* translating to libevmasm assembly.

So *both* ``solc --optimize --bin`` *and* ``solc --optimize --bin --import-json ...`` (after exporting without ``--optimize``) should run the ``Assembly::optimiseInternal`` optimizations and - if things work correctly - produce identical output. Similarly ``solc --bin`` (without ``--optimize``) should be the same as exporting and then importing as ``solc --bin --import-json``. That's the most basic sanity test for the import, so we should definitely check for those, and if their outputs are not identical, we need at least to figure out why that is (it's *not* due to the yul optimizer).

The more obscure cases are optimizing twice, before export and after import - or optimizing before export, but not after import, vs. not optimizing before export, but optimizing after import. But even for those I'd be interested in the behaviour, since technically they could also result in identical output.